lib/noticed/base.rb

```diff
@@ -110,13 +110,25 @@ def delivery_method_for(name, options)
       end
     end
 
-    # Validates that all params are present
     def validate!
+      validate_params_present!
+      validate_options_of_delivery_methods!
+    end
+
+    # Validates that all params are present
+    def validate_params_present!
       self.class.param_names.each do |param_name|
         if params[param_name].nil?
           raise ValidationError, "#{param_name} is missing."
         end
       end
     end
+
+    def validate_options_of_delivery_methods!
+      delivery_methods.each do |delivery_method|
+        method = delivery_method_for(delivery_method[:name], delivery_method[:options])
+        method.validate!(delivery_method[:options])
```

Comment:
I just noticed, `delivery_method_for` takes options as the second argument. We probably don't need that since it's just returning the class, not instantiating it. That would make this a bit more clear.

What if we did...

```
method = delivery_method_for(delivery_method[:name])
method.validate!(delivery_method[:options])
```

Reply:
@excid3 I am sorry, but I am not sure if I understand.

`delivery_method_for ` currently requires the options incase we specify a custom `class` (in cases like custom delivery method)

```ruby
  deliver_by :discord, class: 'DeliveryMethods::Discord'
```

```ruby
    def delivery_method_for(name, options)
      if options[:class]
        options[:class].constantize
      else
        "Noticed::DeliveryMethods::#{name.to_s.classify}".constantize
      end
    end
```

So without passing in the `options`, we would not be able to figure out the class to be used?

I think the best we can do here is, instead of passing the whole options, we can just pass in `options[:class]`, so like,

```ruby
  method = delivery_method_for(delivery_method[:name], delivery_method[:options][:class])
  method.validate!(delivery_method[:options])
```

Anything I am missing? 🙂

Reply:
Ah yeah, you're right. Ignore me. We may want to revisit that in the future and refactor.

Reply:
@excid3 no worries, thanks for clarifying 🙂